internal/filter/lexer.go

```diff
@@ -0,0 +1,149 @@
+//go:generate goyacc -v parser.output -o parser.go parser.y
+
+package filter
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+	"text/scanner"
+)
+
+// regex contains a compiled regexp and is used by the Lexer to match filter identifiers.
+// Currently, it allows to match any character except these inside the curly braces.
+var regex = regexp.MustCompile("[^!&|~<>=()]")
+
+// Parse wraps the auto generated yyParse function.
+// It parses the given filter string and returns on success a Filter instance.
+func Parse(expr string) (rule Filter, err error) {
+	lex := new(Lexer)
+	lex.IsIdentRune = isIdentRune
+	lex.Init(strings.NewReader(expr))
+
+	// scanner.Init sets the error function to nil, therefore, we have to register
+	// our error function after the scanner initialization.
+	lex.Scanner.Error = lex.ScanError
+
+	// Enable parsers error verbose to get more context of the parsing failures
+	yyErrorVerbose = true
+
+	defer func() {
+		// All the grammar rules panics when encountering any errors while reducing the filter rules, so try
+		// to recover from it and return an error instead. Since we're used a named return values, we can set
+		// the err value even in deferred function. See https://go.dev/blog/defer-panic-and-recover
+		if r := recover(); r != nil {
+			lex.err = errors.New(fmt.Sprint(r))
+		}
+
+		err = lex.err
```

Comment:
In case of an error, the `lex.rule` should be cleared as otherwise an incomplete result will be returned.
For example, see this testing output of an invalid expression of mine:

```
    parser_test.go:188: 
        	Error Trace:	/home/apenning/git/github.com/Icinga/icinga-notifications/internal/filter/parser_test.go:188
        	Error:      	Expected nil, but got: &errors.errorString{s:"1:120 (119): syntax error: unexpected T_EQUAL, expecting \")\""}
        	Test:       	TestParser/ParseChain
        	Messages:   	There should be no errors but got: 1:120 (119): syntax error: unexpected T_EQUAL, expecting ")"
    parser_test.go:190: 
        	Error Trace:	/home/apenning/git/github.com/Icinga/icinga-notifications/internal/filter/parser_test.go:190
        	Error:      	Not equal: 
        	            	expected: *filter.Condition(&filter.Condition{op:"=", column:"foo", value:"bar"})
        	            	actual  : *filter.Exists(&filter.Exists{column:"foo"})
        	Test:       	TestParser/ParseChain
```